tpl_lib/dxf/dxf.tpl

```diff
@@ -149,56 +149,148 @@ function bounds_center(bounds) {
 }
 
 
-function quad_bezier(p, t) {
-  var c = [sqr(1 - t), 2 * (1 - t) * t, sqr(t)];
+// Find the spline segment for the t value
+function segment_get_index(spl, knots, t) {
+  // TODO Can this be unified?
+  if (spl.isClosed) {
+    for (let segment = 0; segment < knots.length; segment++)
+      if (t < knots[segment]) return segment - 1
 
-  return {
-    x: c[0] * p[0].x + c[1] * p[1].x + c[2] * p[2].x,
-    y: c[0] * p[0].y + c[1] * p[1].y + c[2] * p[2].y
+  } else
+    for (let segment = spl.degree; segment < knots.length - 1; segment++)
+      if (knots[segment] <= t && t <= knots[segment + 1])
+        return segment
+
+  throw new Error('segment not found');
+}
+
+
+function nurbs_build_verts(spl) {
+  var degree = spl.degree
+  var npts   = spl.ctrlPts.length
+
+  if (degree < 1) throw new Error('degree must be at least 1 (linear)')
+  if (npts - 1 < degree)
+    throw new Error('degree must be less than or equal to point count - 1')
+
+  // Convert control points to homogeneous coordinates
+  var verts = []
+  let nverts = spl.isClosed ? npts + degree * 2 : npts
+
+  for (let i = 0; i < nverts; i++) {
+    let j = spl.isClosed ? (i - degree + npts) % npts : i
+    let weight = spl.isRational ? spl.ctrlPts[j].w : 1
+
+    verts.push([
+      spl.ctrlPts[j].x * weight,
+      spl.ctrlPts[j].y * weight,
+      spl.ctrlPts[j].z * weight,
+      weight
+    ])
   }
+
+  return verts
 }
 
 
-function quad_bezier_length(p) {
-  var l = 0;
-  var v = p[0];
+function nurbs_build_knots(spl) {
+  var degree = spl.degree
+  var npts   = spl.ctrlPts.length
+
+  if (spl.isClosed) {
+    // TODO Is it correct to rewrite the knots values?
+    npts += degree + 1
+
+    var delta = 1.0 / (npts + degree + 1)
+    var knot  = 0
+    var knots = []
+
+    for (let i = 0; i < npts + degree; i++) {
+      knots.push(knot)
+      knot += delta
+    }
 
-  for (var i = 1; i <= 100; i++) {
-    var u = quad_bezier(p, 0.01 * i);
-    l += distance2D(v, u);
-    v = u;
+    knots.push(1.0)
```

Comment:
I'm confused by this code.  If the number of control points is ``n + 1`` and the degree is ``p`` then by the end of this code the number of knots is ``n + 2 * p + 2``  but [bspline-curve-closed](https://pages.mtu.edu/~shene/COURSES/cs3621/NOTES/spline/B-spline/bspline-curve-closed.html) says the number of knots should be ``m + 1`` where ``m = n + p + 1``.  In other words, the number of knots should be ``n + p`` so it's ``p + 2`` too many.

Note, I did modify this code a bit but I think the calculation is the same.  Also, if you correct this code it breaks the spline evaluation, so something else must be wrong too.

Reply:
degree = p; @ 197
npts = n+1; @ 198
npts = n+1+p+1; @ 202

npts + degree = n+1+p+1+p; @ 208
m+1 = n+1+p+1+p
m = n+1+p+p.

So yes, it looks like there are p extra knots...
I wonder if line 202 should be 'npts += 1' instead. Is that the correction you tried?

Reply:
I checked the code that I committed and it looks like I had:
```
function nurbs_build_knots(spl) {
  var degree = spl.degree;
  var npts = spl.ctrlPts.length;
  var knots = [];

  if(spl.isClosed) {
      /* TODO: is it right to rewrite the knots values? */
      npts = npts+degree*2;
      var delta = 1.0/(npts+degree+1);
      var knot = 0.0;
      for(var i=0; i<npts+degree; i++) {
        knots.push(knot);
        knot+=delta;
      }
      knots.push(1.0);
  } else {
```
So I had a loop of npts+degree=n+1+p*2+p, so m+1=n+p+1+p*2 which had p*2 too many knots, according to that page. I remember that I needed that many knots to allow nurbs_interpolate to work more or less in the same way for open and closed nurbs.

I just tried to remove that `npts = npts+degree*2` and modify the nurbs_interpolate accordingly to take into consideration the fewer knots, but something is not right with the min/max calculations for the `t = min + (max-min) * t;` range of closed nurbs: it looks like `var max = knots[knots.length-degree+1]` draws too much on one end, but `var max = knots[knots.length-degree]` too much on the other end.

My test:
```
diff --git a/tpl_lib/dxf/dxf.tpl b/tpl_lib/dxf/dxf.tpl
index e8480d1..5104ddc 100644
--- a/tpl_lib/dxf/dxf.tpl
+++ b/tpl_lib/dxf/dxf.tpl
@@ -155,7 +155,7 @@ function get_segment_index(knots, t, degree, is_closed) {
   /* TODO: can this be unified? */
   if(is_closed) {
     for(var segment = 0; segment<knots.length; segment++) {
-      if(knots[segment]>t) return segment-1;
+      if(knots[segment]>=t) return segment-1;
     }
   } else {
     for(var segment = degree; segment<knots.length; segment++) {
@@ -208,7 +208,6 @@ function nurbs_build_knots(spl) {
   var knots = [];
   if(spl.isClosed) {
       /* TODO: is it right to rewrite the knots values? */
-      npts = npts+degree*2;
       var delta = 1.0/(npts+degree+1);
       var knot = 0.0;
       for(var i=0; i<npts+degree; i++) {
@@ -249,8 +248,8 @@ function nurbs_interpolate(t, spl, vertices, knots) {
   }
 
   if(spl.isClosed) {
-    var min = knots[degree+1];
-    var max = knots[knots.length-degree-1];
+    var min = knots[degree-1];
+    var max = knots[knots.length-degree];
     t = min + (max-min) * t;
   } else {
     var min = knots[degree];
@@ -287,10 +286,11 @@ function nurbs_interpolate(t, spl, vertices, knots) {
     }
   }
 
+       segment=(segment+npts)%npts;
   return {
-    x: verts[segment%npts][0] / verts[segment%npts][3],
-    y: verts[segment%npts][1] / verts[segment%npts][3],
-    z: verts[segment%npts][2] / verts[segment%npts][3]
+    x: verts[segment][0] / verts[segment][3],
+    y: verts[segment][1] / verts[segment][3],
+    z: verts[segment][2] / verts[segment][3]
   } 
 }
 
```

Reply:
Yes, I remember now I did reduce the number of knots by ``p``.  I'm still bothered by not knowing why it works correctly with seemingly the wrong number of knots.